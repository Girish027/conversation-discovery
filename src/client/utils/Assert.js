/**
 * 24/7 Customer, Inc. Confidential, Do Not Distribute. This is an
 * unpublished, proprietary work which is fully protected under
 * copyright law. This code may only be used pursuant to a valid
 * license from 24/7 Customer, Inc.
 */

/**
 * Assertions module
 * @module helpers/assert
 */

const RX_EMPTY = /^\s*$/;

/**
 * Callback invoked when an assertion fails.
 * The default implementation throws an exception that includes a message describing the failed assertion.
 * @func _onFailure
 * @private
 * @memberOf module:helpers/assert
 * @param {String} msg Message to be shown when exception is thrown
 */
function onFailure(msg) {
  throw new Error(msg);
}

class Assert {
  /**
   * Assert a condition and throw an error if the condition is equivalent to `false`.
   * @func assert
   * @public
   * @memberOf module:helpers/assert
   * @param {Boolean} cond A condition to evaluate.
   * @param {String} [msg] The error message to throw.
   */
  static assert(cond, msg = 'assert failed') {
    if (!cond) {
      onFailure(msg);
    }
  }

  /**
   * Assert the given value to be an object.
   * @func isObject
   * @public
   * @memberOf module:helpers/assert
   * @param {Any} obj An object to evaluate.
   * @param {String} [name] The name of the object.
   */
  static isObject(obj, name) {
    const msg = (name ? `${name} is not an object` : 'object expected');
    Assert.assert(typeof obj === 'object' && obj !== null, msg);
  }

  /**
   * Assert the given value to be undefined or null.
   * @func isUndefinedOrNull
   * @public
   * @memberOf module:helpers/assert
   * @param {Any} obj An object to evaluate.
   * @param {String} [name] The name of the object.
   */
  static isUndefinedOrNull(obj, name) {
    const msg = (name ? `${name} is not undefined or null` : 'undefined or null expected');
    Assert.assert(obj === undefined || obj === null, msg);
  }

  /**
   * Assert the given value to be not undefined or null.
   * @func isNotUndefinedOrNull
   * @public
   * @memberOf module:helpers/assert
   * @param {Any} obj An object to evaluate.
   * @param {String} [name] The name of the object.
   */
  static isNotUndefinedOrNull(obj, name) {
    const msg = (name ? `${name} is undefined or null` : 'undefined or null not expected');
    Assert.assert(obj !== undefined && obj !== null, msg);
  }

  /**
   * Assert the given value to be a boolean.
   * @func isBoolean
   * @public
   * @memberOf module:helpers/assert
   * @param {Any} bool A boolean to evaluate.
   * @param {String} [name] The name of the boolean.
   */
  static isBoolean(bool, name) {
    const msg = (name ? `${name} is not a boolean` : 'boolean expected');
    Assert.assert(typeof bool === 'boolean', msg);
  }

  /**
   * Assert the given value to be a function.
   * @func isFunction
   * @public
   * @memberOf module:helpers/assert
   * @param {Any} func A function to evaluate.
   * @param {String} [name] The name of the function.
   */
  static isFunction(func, name) {
    const msg = (name ? `${name} is not a function` : 'function expected');
    Assert.assert(typeof func === 'function', msg);
  }

  /**
   * Assert the given value to be a number.
   * @func isNumber
   * @public
   * @memberOf module:helpers/assert
   * @param {Any} num A number to validate.
   * @param {String} [name] The name of the variable that stores the number.
   */
  static isNumber(num, name) {
    const msg = (name ? `${name} is not a number` : 'number expected');
    Assert.assert(typeof num === 'number', msg);
  }

  /**
   * Assert the given value is a string and not empty.
   * @func notEmpty
   * @public
   * @memberOf module:helpers/assert
   * @param {String} str A string to evaluate.
   * @param {String} [name] The name of the string.
   */
  static notEmpty(str, name) {
    const msg = (name ? `${name} is not a string or is empty` : 'non-empty string expected');
    Assert.assert(typeof str === 'string' && !RX_EMPTY.test(str), msg);
  }

  /**
   * Assert the given object has the given constructor.
   * @func instanceOf
   * @public
   * @memberOf module:helpers/assert
   * @param {Object} obj An object to evaluate.
   * @param {Object.constructor} constructor A constructor to evaluate.
   * @param {String} [className] The name of the class/constructor.
   */
  static instanceOf(obj, constructor, className) {
    const msg = (className ? `expected instance of ${className}` : 'unexpected instance type');
    Assert.assert(obj instanceof constructor, msg);
  }

  /**
   * Assert the given value to be a non-empty string, a number, or a boolean.
   * @func isScalar
   * @public
   * @memberOf module:helpers/assert
   * @param {Any} value A variant to evaluate.
   * @param {String} [name] The name of the variable that stores the variant.
   */
  static isScalar(value, name) {
    const vType = typeof value;
    const msg = `${name ? `${name} ` : ''}must be string, number, or boolean`;
    Assert.assert(vType === 'string' || vType === 'number' || vType === 'boolean', msg);
    if (vType === 'string') {
      Assert.notEmpty(value, name);
    }
  }

  /**
   * Assert the given value to be an array.
   * @func isArray
   * @public
   * @memberOf module:helpers/assert
   * @param {Any} value A variant to evaluate.
   * @param {String} [name] The name of the variable that stores the variant.
   */
  static isArray(value, name) {
    const msg = (name ? `${name} is not an array` : 'array expected');
    Assert.assert(Object.prototype.toString.call(value) === '[object Array]', msg);
  }
}


export default Assert;
