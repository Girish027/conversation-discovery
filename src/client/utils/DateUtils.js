import dateFormat from 'dateformat';

export const getDateFromTimestamp = (timestamp, locale = 'en-US') => new Date(parseInt(timestamp, 10))
  .toLocaleDateString(locale, {
    day: 'numeric',
    month: 'short',
    year: 'numeric',
  });

export const formatTimestamp = (timestamp, format = 'mmm d, h:MM TT') => {
  const dateObj = new Date(parseInt(timestamp, 10));
  return dateFormat(dateObj, format);
};
