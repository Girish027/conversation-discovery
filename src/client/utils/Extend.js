/**
 * 24/7 Customer, Inc. Confidential, Do Not Distribute. This is an
 * unpublished, proprietary work which is fully protected under
 * copyright law. This code may only be used pursuant to a valid
 * license from 24/7 Customer, Inc.
 */

/**
 * A module that exports an extend function
 * Use this IFF jQuery and underscore are not available.
 * @module utils/extend
 * @see http://andrewdupont.net/2009/08/28/deep-extending-objects-in-javascript/
 * Extend the target object with the given object(s),
 * ala jQuery.extend(), with deep copy support.
 * The function takes a variable argument list of Objects.
 * @function
 * @param {Object} target the target object to extend
 * @param {Object} obj... the object to use to extend the target
 */
const extend = function extend(target, ...args) {
  if (!(target !== undefined && typeof target === 'object')) {
    return target;
  }

  const _target = target;

  const argCount = args.length;
  for (let idx = 0; idx < argCount; idx += 1) {
    const obj = args[idx];
    if (obj !== undefined && typeof obj === 'object') {
      Object.keys(obj).map((prop) => {
        if (Object.prototype.hasOwnProperty.call(obj, prop)) {
          if (obj[prop] !== undefined
            && obj[prop].constructor
            && obj[prop].constructor === Object) {
            if (_target[prop] === undefined) {
              _target[prop] = {};
            }
            if (typeof _target[prop] !== 'object') {
              _target[prop] = obj[prop];
            } else {
              extend(_target[prop], obj[prop]);
            }
          } else if (obj[prop] && obj[prop].constructor === Array) {
            _target[prop] = obj[prop].slice(0);
          } else {
            _target[prop] = obj[prop];
          }
        }
        return 0;
      });
    }
  }
  return _target;
};

export default extend;
