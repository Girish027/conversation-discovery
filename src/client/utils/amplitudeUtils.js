// import amplitude from 'amplitude-js/amplitude';

export const initializeAnalytics = (analyticsKey, userName = '') => {
  // const analytics = amplitude.getInstance();
  // analytics.init(analyticsKey, userName);
};

export const logAmplitudeEvent = (eventName, eventSpecificData = {}) => {
  // amplitude.getInstance().logEvent(eventName, {
  //   ...eventSpecificData,
  // });
};
