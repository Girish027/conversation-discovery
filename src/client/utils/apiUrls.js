import { basePath } from '../swagger.json';

const root = (process.env.NODE_ENV === 'production'
  ? '//'.concat(window.location.host)
  : '//localhost:8080')
  .concat(basePath);

const clientInfoPath = (data) => `${root}/clients/${data.componentClientId}/accounts/${data.accountId}/applications/${data.appId}`;
const projectInfo = ({ projectId }) => `projects/${projectId}`;
// TODO remove mockServer once APIs are implemented
// const runsInfo = ({ runId }) => `runs/${runId}/mockServer/`;
const runsInfo = ({ runId }) => `runs/${runId}`;
const clusterInfo = ({ clusterId }) => `clusters/${clusterId}`;
const transcriptInfo = ({ transcriptId }) => `conversations/${transcriptId}`;
const conversationTranscriptInfo = (transcriptId) => `conversations?tId=${transcriptId}`;

// TODO: Curation
// API: ....conversations?transcriptIds=['abc', 'def']&moveTo=<new cluster id>

export const apiUrls = {
  authentications: `${root}/authentications?self=true`,
  userConfigUrl: `${root}/user/config`,
  conversationsConfigUrl: `${root}/user/conversationsConfig`,
  answersConfigUrl: `${root}/user/answersConfig`,
  downloadTemplate: (data) => `${clientInfoPath(data)}/templates`,
  projects: (data) => `${clientInfoPath(data)}/projects`,
  oneProject: (data) => `${clientInfoPath(data)}/${projectInfo(data)}`,
  runs: (data) => `${clientInfoPath(data)}/${projectInfo(data)}/runs`,
  oneRun: (data) => `${clientInfoPath(data)}/${projectInfo(data)}/${runsInfo(data)}`,
  downloadRun: (data) => `${clientInfoPath(data)}/${projectInfo(data)}/${runsInfo(data)}/results?type=original`,
  clusters: (data) => `${clientInfoPath(data)}/${projectInfo(data)}/${runsInfo(data)}/clusters`,
  oneCluster: (data) => `${clientInfoPath(data)}/${projectInfo(data)}/${runsInfo(data)}/${clusterInfo(data)}`,
  conversations: (data) => `${clientInfoPath(data)}/${projectInfo(data)}/${runsInfo(data)}/${clusterInfo(data)}/conversations`,
  transcript: (data) => `${clientInfoPath(data)}/${projectInfo(data)}/${runsInfo(data)}/${clusterInfo(data)}/${transcriptInfo(data)}`,
  intents: (data) => `${clientInfoPath(data)}/${projectInfo(data)}/runs/${data.runId}/${clusterInfo(data)}/intents`,
  conversationTranscript: (data, transcriptId) => `${clientInfoPath(data)}/${projectInfo(data)}/${runsInfo(data)}/${clusterInfo(data)}/${conversationTranscriptInfo(transcriptId)}`,
  interfaces: (data) => `${root}/clients/${data.componentClientId}/interfaces`,
  faqs: (data, interfaceId) => `${clientInfoPath(data)}/${projectInfo(data)}/${runsInfo(data)}/${clusterInfo(data)}/interfaceId/${interfaceId}/addToAnswers`
};
