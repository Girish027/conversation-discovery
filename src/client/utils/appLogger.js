// log to console if method level meets log level
function appLogger(options) {
  // indexes of method names match values of levels properties
  const loggerMethods = ['trace', 'debug', 'info', 'warn', 'error'];
  const levels = {
    TRACE: 0,
    DEBUG: 1,
    INFO: 2,
    WARN: 3,
    ERROR: 4,
    SILENT: 5
  };
  const envLevels = {
    development: levels.TRACE,
    production: levels.SILENT,
    test: levels.ERROR
  };

  // environment variables can override default log level
  let logLevel = levels.INFO;
  if (typeof process.env.LOG_LEVEL !== 'undefined') {
    logLevel = levels[process.env.LOG_LEVEL];
  } else if (typeof process.env.NODE_ENV !== 'undefined') {
    logLevel = envLevels[process.env.NODE_ENV];
  }

  const logger = {};
  loggerMethods.forEach((methodName, methodLevel) => {
    logger[methodName] = (...args) => {
      // only print out if method level is not less than log level
      if (methodLevel >= logLevel) {
        let msgs = [];
        if (options && options.tag) {
          msgs.push(options.tag);
        }
        msgs = msgs.concat(args);
        /* eslint-disable no-alert, no-console */
        console.log(...msgs);
      }
    };
  });
  return logger;
}

export default appLogger;
