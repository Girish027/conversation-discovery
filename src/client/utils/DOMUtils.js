import Assert from './Assert';
import TimerUtils from './TimerUtils';
import Extend from './Extend';
import ObjectUtils from './ObjectUtils';

class DOMUtils {
  trim(str) {
    return str.replace(/^\s+|\s+$/g, '');
  }

  addStylesheet(id, cssText) {
    Assert.notEmpty(id, 'addStylesheet id');
    Assert.notEmpty(cssText, 'addStylesheet cssText');

    const el = this.createElement('style', {
      id,
      type: 'text/css'
    });

    try {
      if (el.styleSheet) {
        el.styleSheet.cssText = cssText;
      } else {
        el.appendChild(document.createTextNode(cssText));
      }
      document.head.appendChild(el);
    } catch (e) {
      console.log('DOMUtils.addStylesheet error: ', e);
    }
    return el;
  }

  get(id) {
    return document.getElementById(id);
  }

  createElement(tag, attr = {}, children = []) {
    const el = document.createElement(tag);
    Object.keys(attr).map((key) => {
      el.setAttribute(key, attr[key]);
      return 0;
    });

    if (typeof children === 'string') {
      el.innerHTML = children;
    } else {
      children.map((child) => el.appendChild(child));
    }
    return el;
  }

  removeElement(el) {
    let _el = el;
    if (typeof _el === 'string') {
      _el = this.get(_el);
    }
    if (_el) {
      _el.parentNode.removeChild(_el);
    }
  }

  addClass(el, name) {
    const _name = this.trim(name);
    let _el = el;
    if (!_el) {
      return;
    }
    if (typeof _el === 'string') {
      _el = this.get(_el);
    }
    if (_el.classList) {
      _el.classList.add(_name);
    } else {
      _el.className += ` ${_name}`;
    }
  }

  removeClass(el, name) {
    const _name = this.trim(name);
    let _el = el;
    if (!_el) {
      return;
    }
    if (typeof _el === 'string') {
      _el = this.get(_el);
    }
    if (_el.classList) {
      _el.classList.remove(name);
    } else {
      _el.className = _el.className.replace(new RegExp(`(^|\\b)(\\s*)${_name.split(' ').join('(\\s*)|(\\s*)')}(\\s*)(\\b|$)`, 'g'), ' ');
    }
  }

  hasClass(el, name) {
    const _name = this.trim(name);
    let _el = el;
    if (typeof _el === 'string') {
      _el = this.get(_el);
    }
    if (_el.classList) {
      return _el.classList.contains(_name);
    }
    return new RegExp(`(^| )${_name}( |$)`, 'gi').test(_el.className);
  }

  hasSelector(elContent, selector) {
    const elem = (typeof elContent === 'string') ? this.createElement('div', undefined, elContent) : elContent;
    const result = elem.querySelectorAll(selector);

    return result.length > 0;
  }

  hide(el) {
    let _el = el;
    if (typeof _el === 'string') {
      _el = this.get(_el);
    }
    _el.style.display = 'none';
  }

  show(el) {
    let _el = el;
    if (typeof _el === 'string') {
      _el = this.get(_el);
    }
    _el.style.display = 'block';
  }

  getStyle(element, property) {
    return window.getComputedStyle(element)[property];
  }

  set(el, prop, value) {
    const _el = el;
    if (_el && typeof prop === 'string' && typeof value === 'string') {
      _el.style[prop] = value;
    }
  }

  /**
   * Sets css style for the provided element
   * @param {HTMLElement} element Element for which style has to be set
   * @param {(Object|String)} style CSS properties provided as cssText or key-value pairs object
   * @param {Boolean} [shouldClear=false] Reset existing inline styles before applying new
   * @returns {undefined}
   */
  setStyle(element, style, shouldClear = false) {
    if (!element) {
      return;
    }
    if (shouldClear) {
      element.setAttribute('style', '');
    }
    switch (typeof style) {
      case 'object':
        Object.keys(style).map((prop) => {
          if (Object.prototype.hasOwnProperty.call(style, prop)) {
            this.set(element, prop, style[prop]);
          }
          return 0;
        });
        break;
      case 'string':
        element.setAttribute('style', element.style.cssText + style);
        break;
      default:
        break;
    }
  }

  injectScript(url, onload = () => {}, doc, attr, timeout) {
    Assert.notEmpty(url, 'injectScript url');
    Assert.isNotUndefinedOrNull(doc.head, 'document.head');

    let done = false;
    if (!this.get(attr.id)) {
      console.log('Dynamically adding the script with source : ', url);
      const script = this.createElement('script', Extend({}, attr, {
        async: true,
        src: url,
        type: 'text/javascript'
      }));
      doc.head.appendChild(script);
      let tid;
      script.onreadystatechange = () => {
        if (!done && (!this.readyState || this.readyState === 'loaded' || this.readyState === 'complete')) {
          done = true;
          if (tid) {
            TimerUtils.clearRequestTimeout(tid);
          }
          onload('success');

          // Handle memory leak in IE
          script.onreadystatechange = null;
        }
      };
      script.onload = () => {
        if (!done && (!this.readyState || this.readyState === 'loaded' || this.readyState === 'complete')) {
          done = true;
          if (tid) {
            TimerUtils.clearRequestTimeout(tid);
          }
          onload('success');

          // Handle memory leak in IE
          script.onload = null;
        }
      };
      script.onerror = () => {
        script.onreadystatechange = null;
        script.onload = null;
        if (tid) {
          TimerUtils.clearRequestTimeout(tid);
        }
        onload('error');
      };
      if (timeout) {
        tid = TimerUtils.requestTimeout(() => {
          script.onreadystatechange = null;
          script.onload = null;
          onload('timeout');
        }, timeout);
      }
    } else {
      // script exists
      console.log('Script exists: ', url);
      onload('success');
    }
  }

  /**
   * reduceTags
   * @param  {string} string Input string
   * @param  {Array} allowedTags Allowed tags array
   * @param  {Object} allowedAttributes Allowed attributes map of tag to attributes array
   * @param  {Array} specialTags Tags which need to be flushed completely
   * @return {string} Reduced tags html string
   */
  reduceTags(string, allowedTags = [], allowedAttributes = {}, purgeTags = []) {
    Assert.isArray(allowedTags, 'reduceTags allowedTags');
    Assert.isObject(allowedAttributes, 'reduceTags allowedAttributes');
    Assert.isArray(purgeTags, 'reduceTags purgeTags');

    const dummyElement = this.createElement('div', {}, string);
    const childCollection = dummyElement.childNodes;
    const filteredString = [];

    if (childCollection.length > 0) {
      for (let i = 0; i < childCollection.length; i += 1) {
        let node = childCollection[i];
        const nodeName = node.nodeName.toLowerCase();
        node = !Array.isArray(allowedAttributes[nodeName])
          ? this.removeAttrs(node, [])
          : this.removeAttrs(node, allowedAttributes[nodeName]);
        if (node.hasChildNodes() && purgeTags.indexOf(nodeName) === -1) {
          if (node.childNodes.length === 1 && node.childNodes[0].nodeName === '#text') {
            if (allowedTags.indexOf(nodeName) > -1) {
              filteredString.push(node.outerHTML);
            } else {
              filteredString.push(node.textContent);
            }
          } else {
            const tagPositionOpen = node.innerHTML.indexOf('<');
            let partialPreText = '';
            if (tagPositionOpen > 0) {
              partialPreText = node.innerHTML.substr(0, tagPositionOpen);
              node.innerHTML = node.innerHTML.substr(tagPositionOpen);
            }
            const tagPositionClose = node.innerHTML.lastIndexOf('>');
            let partialPostText = '';
            partialPostText = node.innerHTML.substr(tagPositionClose + 1);
            node.innerHTML = node.innerHTML.substr(0, tagPositionClose + 1);
            if (allowedTags.indexOf(nodeName) > -1) {
              node.innerHTML = partialPreText + this.reduceTags(node.innerHTML, allowedTags, allowedAttributes, purgeTags) + partialPostText;
              filteredString.push(node.outerHTML);
            } else {
              filteredString.push(partialPreText + this.reduceTags(node.innerHTML, allowedTags, allowedAttributes, purgeTags) + partialPostText);
            }
          }
        } else {
          const currentTag = nodeName;
          if (purgeTags.indexOf(currentTag) === -1) {
            if (allowedTags.indexOf(currentTag) > -1) {
              filteredString.push(node.outerHTML);
            } else {
              filteredString.push(node.textContent);
            }
          }
        }
      }
    } else {
      filteredString.push(string);
    }

    return filteredString.join(' ').trim();
  }

  removeAttrs(element, allowedAttributes) {
    const el = element;
    if (el.attributes) {
      const attrs = el.attributes;
      for (let i = attrs.length - 1; i >= 0; i -= 1) {
        if (allowedAttributes.indexOf(attrs[i].name) === -1) {
          el.removeAttribute(attrs[i].name);
        }
      }
    }
    return el;
  }

  removeCSSProp(styleStr, allowedProps) {
    const str = styleStr;
    const matches = str.match(/([^\s{]+?(?=:))(:\s)([^;]*);/g);
    let cleanStr = '';
    if (matches) {
      matches.forEach((item) => {
        const prop = item.split(':')[0];
        if (allowedProps.indexOf(prop) > -1) {
          cleanStr += item;
        }
      });
    }
    return cleanStr;
  }

  fixAnchorTarget(string) {
    const dummyEl = this.createElement('div', {}, string);
    const elList = dummyEl.getElementsByTagName('a');
    let filteredString;

    if (elList.length > 0) {
      for (let i = 0; i < elList.length; i += 1) {
        const target = elList[i].getAttribute('target');
        if (ObjectUtils.isEmptyOrNull(target)) {
          elList[i].setAttribute('target', '_blank');
        }
      }
      filteredString = dummyEl.innerHTML;
    } else {
      filteredString = string;
    }

    return filteredString;
  }

  filterAttrOfTag(string, tag, attributes) {
    const dummyElement = this.createElement('div', {}, string);
    const elementCollection = dummyElement.getElementsByTagName(tag);
    const tagAttributes = [];
    let residueContent;

    const elementArray = [].slice.call(elementCollection);

    if (elementArray.length === 0) {
      residueContent = dummyElement.innerHTML;
      return { residueContent };
    }

    elementArray.map((element) => {
      const attributeObj = {};

      attributes.map((attribute) => {
        const elementValue = (!ObjectUtils.isEmptyOrNull(element.getAttribute(attribute)))
          ? element.getAttribute(attribute).trim()
          : element.getAttribute(attribute);
        attributeObj[attribute] = (!ObjectUtils.isEmptyOrNull(element.dataset)
          && !ObjectUtils.isEmptyOrNull(element.dataset[attribute]))
          ? element.dataset[attribute].trim()
          : elementValue;
        return 0;
      });
      attributeObj.innerHTML = element.innerHTML.trim();

      tagAttributes.push(attributeObj);
      this.removeElement(element);
      return 0;
    });

    if ((tagAttributes.length > 0
      && dummyElement.textContent.trim() === '')
      && dummyElement.getElementsByTagName('img').length === 0) {
      residueContent = undefined;
    } else {
      residueContent = dummyElement.innerHTML;
    }

    return { residueContent, tagAttributes };
  }

  redrawElement(element, style1, style2) {
    if (!element) {
      return 0;
    }
    this.setStyle(element, style1);
    TimerUtils.requestTimeout(() => {
      this.setStyle(element, style2);
    }, 40);
    return 0;
  }

  matches(el, selector) {
    const matches = (el.document || el.ownerDocument).querySelectorAll(selector);
    let i = matches.length;
    while (i >= 0 && matches.item(i) !== el) {
      i -= 1;
    }
    return i > -1;
  }
}

export default new DOMUtils();
