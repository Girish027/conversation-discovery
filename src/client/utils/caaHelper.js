import appLogger from 'utils/appLogger';
import memoizeOne from 'memoize-one';

const devlog = appLogger({ tag: 'caaHelper' });

export const getSortedClientsList = (clientsList = []) => clientsList
  .map((config) => ([config.clientId.toLowerCase(), config]))
// eslint-disable-next-line no-nested-ternary
  .sort((a, b) => (a[0] > b[0] ? 1 : a[0] < b[0] ? -1 : 0))
  .map(([, config]) => config);


export const getDefaultCAA = (sortedClientsList) => {
  const { clientId, apps = [{}], componentClientId } = sortedClientsList[0];
  const { appId, accountId } = apps[0];
  return {
    clientId,
    appId,
    accountId,
    componentClientId,
  };
};

export const getMatchingAccountId = memoizeOne((clientConfigs, clientId, appId) => {
  const matchingClient = clientConfigs.find((cl) => cl.clientId === clientId) || {};
  const { apps = [] } = matchingClient;
  const matchingApp = apps.find((app) => app.appId === appId) || {};
  const { accountId = 'default' } = matchingApp;
  return accountId;
});

export const getComponentClientIdForClient = memoizeOne((clientConfigs, clientId) => {
  const matchingClient = clientConfigs.find((cl) => cl.clientId === clientId) || {};
  return matchingClient.componentClientId;
});

export const isUserAuthorized = (userSelectedCAA = {}, clientsList = [{}]) => {
  try {
    const {
      clientId: userClientId,
      appId: userAppId,
      accountId: userAccountId
    } = userSelectedCAA;
    // caa information is neither present in URL or cookie
    // user is authorized to view default CAA
    if (userClientId === undefined || userClientId === '') {
      if (clientsList.length === 1 && clientsList[0].apps.length === 1) {
        return {
          authorized: true,
          clientPicker: false,
          caa: getDefaultCAA(clientsList)
        };
      }
      return {
        authorized: true,
        clientPicker: true,
        caa: {}
      };
    }

    // caa information is present in URL.
    const clientConfig = clientsList.find((client) => client.clientId === userClientId);
    if (clientConfig) {
      if (userAppId === undefined && userAccountId === '') {
        if (clientConfig.apps.length === 1) {
          return {
            authorized: true,
            clientPicker: false,
            caa: {
              clientId: userClientId,
              appId: clientConfig.apps[0].appId,
              accountId: clientConfig.apps[0].accountId,
              componentClientId: clientConfig.componentClientId,
            }
          };
        }
        return {
          authorized: true,
          clientPicker: true,
          caa: {}
        };
      }

      let appDetails = clientConfig.apps.find((app) => app.appId === userAppId);
      if (!appDetails) {
        const [first] = clientConfig.apps;
        if (first) {
          appDetails = first;
        }
      }
      // we were able to find matching client and app.
      const { accountId, appId } = appDetails;
      userSelectedCAA.appId = appId;
      userSelectedCAA.accountId = accountId;
      userSelectedCAA.componentClientId = clientConfig.componentClientId;

      return {
        authorized: true,
        clientPicker: false,
        caa: userSelectedCAA
      };
    }
  } catch (err) {
    devlog.warn(err);
  }
  // return false and defaultCAA if we were not able to find matching CAA in clientsList
  return {
    authorized: false,
    clientPicker: false,
    caa: getDefaultCAA(clientsList)
  };
};
