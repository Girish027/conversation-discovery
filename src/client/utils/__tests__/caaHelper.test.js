import * as caaHelper from 'utils/caaHelper';

describe('utils/caaHelper', function () {
  const clientsList = [{
    clientId: 'capitalone',
    apps: [{
      appId: 'chatbot',
      accountId: 'chatbot'
    }]
  }, {
    clientId: 'sirius',
    apps: [{
      appId: 'chatbot',
      accountId: 'chatbot'
    }]
  }, {
    clientId: 'ebay',
    apps: [{
      appId: 'uk',
      accountId: 'uk'
    }, {
      appId: 'de',
      accountId: 'de'
    }]
  }, {
    clientId: 'assurant',
    apps: [{
      appId: 'pyb',
      accountId: 'pyb'
    }]
  }];
  const sortedClientsList = [{
    clientId: 'assurant',
    apps: [{
      appId: 'pyb',
      accountId: 'pyb'
    }],
    componentClientId: 'assurant'
  }, {
    clientId: 'capitalone',
    apps: [{
      appId: 'chatbot',
      accountId: 'chatbot'
    }],
    componentClientId: 'capitalone'
  }, {
    clientId: 'ebay',
    apps: [{
      appId: 'uk',
      accountId: 'uk'
    }, {
      appId: 'de',
      accountId: 'de'
    }],
    componentClientId: 'ebay'
  }, {
    clientId: 'sirius',
    apps: [{
      appId: 'chatbot',
      accountId: 'chatbot'
    }],
    componentClientId: 'sirius'
  }];

  describe('getSortedClientsList', () => {
    test('should sort the clientsList', () => {
      expect(caaHelper.getSortedClientsList(clientsList)).toMatchSnapshot();
    });

    test('should sort the clientsList', () => {
      expect(caaHelper.getSortedClientsList()).toMatchSnapshot();
    });
  });

  describe('getDefaultCAA', () => {
    test('should get the default caa - first client, app, account', () => {
      expect(caaHelper.getDefaultCAA(sortedClientsList)).toMatchSnapshot();
    });

    const clientsList = [{
      clientId: 'assurant',
      componentClientId: 'assurant'
    }];
    test('should get the default caa - first client, app, account', () => {
      expect(caaHelper.getDefaultCAA(clientsList)).toMatchSnapshot();
    });
  });

  describe('getMatchingAccountId', () => {
    test('should get the default caa - first client, app, account', () => {
      expect(caaHelper.getMatchingAccountId(sortedClientsList, 'sirius', 'chatbot')).toEqual('chatbot');
    });
  });

  describe('isUserAuthorized', () => {
    test('should authorize and get the default caa when  CAA is not provided', () => {
      expect(caaHelper.isUserAuthorized(undefined, sortedClientsList)).toEqual({
        authorized: true,
        clientPicker: true,
        caa: {}
      });
    });

    test('should authorize and get the provided caa when user is authorized', () => {
      const userSelectedCAA = {
        clientId: 'sirius',
        appId: 'chatbot',
        accountId: 'chatbot',
      };
      expect(caaHelper.isUserAuthorized(userSelectedCAA, sortedClientsList)).toEqual({
        authorized: true,
        clientPicker: false,
        caa: userSelectedCAA
      });
    });

    test('should not authorize and get the default caa when user is not authorized', () => {
      const userSelectedCAA = {
        clientId: 'differentClient',
        appId: 'chatbot',
        accountId: 'chatbot',
      };
      expect(caaHelper.isUserAuthorized(userSelectedCAA, sortedClientsList)).toEqual({
        authorized: false,
        clientPicker: false,
        caa: caaHelper.getDefaultCAA(sortedClientsList)
      });
    });
  });
});
