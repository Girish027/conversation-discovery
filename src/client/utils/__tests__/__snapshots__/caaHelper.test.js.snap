// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/caaHelper getDefaultCAA should get the default caa - first client, app, account 1`] = `
Object {
  "accountId": "pyb",
  "appId": "pyb",
  "clientId": "assurant",
  "componentClientId": "assurant",
}
`;

exports[`utils/caaHelper getDefaultCAA should get the default caa - first client, app, account 2`] = `
Object {
  "accountId": undefined,
  "appId": undefined,
  "clientId": "assurant",
  "componentClientId": "assurant",
}
`;

exports[`utils/caaHelper getSortedClientsList should sort the clientsList 1`] = `
Array [
  Object {
    "apps": Array [
      Object {
        "accountId": "pyb",
        "appId": "pyb",
      },
    ],
    "clientId": "assurant",
  },
  Object {
    "apps": Array [
      Object {
        "accountId": "chatbot",
        "appId": "chatbot",
      },
    ],
    "clientId": "capitalone",
  },
  Object {
    "apps": Array [
      Object {
        "accountId": "uk",
        "appId": "uk",
      },
      Object {
        "accountId": "de",
        "appId": "de",
      },
    ],
    "clientId": "ebay",
  },
  Object {
    "apps": Array [
      Object {
        "accountId": "chatbot",
        "appId": "chatbot",
      },
    ],
    "clientId": "sirius",
  },
]
`;

exports[`utils/caaHelper getSortedClientsList should sort the clientsList 2`] = `Array []`;
